import os
import pandas as pd
from os.path import join
from HBVouroboros import biokit
import snakemake


configfile: snakemake.workflow.srcdir("config/config.yaml")

create_genome_size_cmd='singularity exec -B /pstore/data/:/pstore/data/:rw /pstore/data/biomics/ID/7907_HBV_Capsid_2/202006-plasmid-validation/test-pisces/pisces-s3.5.simg dotnet /app/CreateGenomeSizeFile_5.2.9.122/CreateGenomeSizeFile.dll'
pisces_cmd='singularity exec -B /pstore/data/:/pstore/data/:rw /pstore/data/biomics/ID/7907_HBV_Capsid_2/202006-plasmid-validation/test-pisces/pisces-s3.5.simg dotnet /app/Pisces_5.2.9.122/Pisces.dll' 

sample_annotation = config['sample_annotation']
input_fa = config['ref_genome_fa']
ref_genome_str = '{genus} {species} ({build})'.format(
    genus=config['ref_genome_genus'],
    species=config['ref_genome_species'],
    build=config['ref_genome_build'])
    
samples, fq1dict, fq2dict = biokit.parse_sample_annotation(sample_annotation)

ref_genome_dir = "reference_genome"
rawbam_dir = "raw_bam"
bam_dir = "bam"
pisces_dir = "pisces"
log_dir = "logs"
stats_dir = "stats"

include: '../align_reads/map-fastq-snakefile'

ref_genome_GenomeFaFile = join(ref_genome_dir, 'genome.fa')
ref_genome_GenomeSizeFile = join(ref_genome_dir, 'GenomeSize.xml')

rule all:
    input:
        ref_genome_GenomeSizeFile

rule reference_genome_dir:
    output:
        gf=directory(ref_genome_dir)
    shell:
        "mkdir -p {reference_genome_dir}"

rule reference_genome_fa:
    input:
        input_fa
    output:
        ref_genome_GenomeFaFile,
    shell:
        "cp -pr {input} {output}"
        
rule reference_genome_gs:
    input:
        fa=ref_genome_GenomeFaFile
    output:
        gs=ref_genome_GenomeSizeFile
    shell:
        "{create_genome_size_cmd} -g {ref_genome_dir} "
        "-s \"{ref_genome_str}\" -o {ref_genome_dir}"

rule pisces:
    input:
        bam=join(bam_dir, "{sample}.sorted.bam"),
        gs=ref_genome_GenomeSizeFile
    output:
        directory(join(pisces_dir, "{sample}"))
    shell:
        "{pisces_cmd} --bam {input} -g {ref_genome_dir} -o {output}"
